cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/analyzer/ProvisionResponseTest.java

```diff
@@ -13,72 +15,179 @@
+    assertEquals(String.format("[%s] %s", recommenderUp, ProvisionResponse.DEFAULT_RECOMMENDATION), nullRecommendationUp.recommendation());
+    assertEquals(String.format("[%s] %s", recommenderOp, ProvisionResponse.DEFAULT_RECOMMENDATION), nullRecommendationOp.recommendation());
+
+    // Verify recommendationByRecommender is empty
+    assertTrue(nullRecommendationUp.recommendationByRecommender().isEmpty());
+    assertTrue(nullRecommendationOp.recommendationByRecommender().isEmpty());
+
+    // Verify aggregating with another empty recommendation with the same state.
+    String recommenderUp2 = "nullRecommendationUp2";
+    String recommenderOp2 = "nullRecommendationOp2";
+    ProvisionResponse nullRecommendationUp2 = new ProvisionResponse(ProvisionStatus.UNDER_PROVISIONED, null, recommenderUp2);
+    ProvisionResponse nullRecommendationOp2 = new ProvisionResponse(ProvisionStatus.OVER_PROVISIONED, null, recommenderOp2);
+
+    nullRecommendationUp.aggregate(nullRecommendationUp2);
+    nullRecommendationOp.aggregate(nullRecommendationOp2);
+
+    assertEquals(String.format("[%s] %s [%s] %s", recommenderUp, ProvisionResponse.DEFAULT_RECOMMENDATION, recommenderUp2,
+                               ProvisionResponse.DEFAULT_RECOMMENDATION), nullRecommendationUp.recommendation());
+    assertEquals(String.format("[%s] %s [%s] %s", recommenderOp, ProvisionResponse.DEFAULT_RECOMMENDATION, recommenderOp2,
+                               ProvisionResponse.DEFAULT_RECOMMENDATION), nullRecommendationOp.recommendation());
+
+    // Verify recommendationByRecommender is still empty
+    assertTrue(nullRecommendationUp.recommendationByRecommender().isEmpty());
+    assertTrue(nullRecommendationOp.recommendationByRecommender().isEmpty());
+
+    // Verify aggregating with a non-empty recommendation with the same state.
+    nullRecommendationUp.aggregate(generateProvisionResponse(ProvisionStatus.UNDER_PROVISIONED));
+    nullRecommendationOp.aggregate(generateProvisionResponse(ProvisionStatus.OVER_PROVISIONED));
+
+    assertEquals(String.format("[%s] %s [%s] %s [%s] %s", recommenderUp, ProvisionResponse.DEFAULT_RECOMMENDATION, recommenderUp2,
+                               ProvisionResponse.DEFAULT_RECOMMENDATION, RECOMMENDER_UP, UNDER_PROV_REC_STR), nullRecommendationUp.recommendation());
+    assertEquals(String.format("[%s] %s [%s] %s [%s] %s", recommenderOp, ProvisionResponse.DEFAULT_RECOMMENDATION, recommenderOp2,
+                               ProvisionResponse.DEFAULT_RECOMMENDATION, RECOMMENDER_OP, OVER_PROV_REC_STR), nullRecommendationOp.recommendation());
+
+    // Verify recommendationByRecommender is no longer empty and has the expected value
+    assertEquals(1, nullRecommendationUp.recommendationByRecommender().size());
+    assertEquals(1, nullRecommendationOp.recommendationByRecommender().size());
+    assertEquals(NUM_BROKERS_UP, nullRecommendationUp.recommendationByRecommender().get(RECOMMENDER_UP).numBrokers());
```

Comment:
Is it still relevant to verify `OVER_PROVISIONED` scenario here? Just wondering because we omit the counter part and verify only the `UNDER_PROVISIONED` from here on.

Reply:
Absolutely, thanks for pointing this out!